telemetry/data_slow_query.go

```diff
@@ -0,0 +1,187 @@
+	defer func() {
+		if r := recover(); r != nil {
+			switch x := r.(type) {
+			case string:
+				err = errors.New(x)
+			case error:
+				err = x
+			default:
+				err = errors.New("unknown failure")
+			}
+		}
+	}()
+
+	value, err := querySlowQueryMetric(ctx) //TODO: judge error here
+	if err != nil && err != infosync.ErrPrometheusAddrIsNotSet {
+		logutil.BgLogger().Info("querySlowQueryMetric got error")
+		return err
+	}
+
+	if value.Type() != pmodel.ValVector {
+		return errors.New("Prom vector expected, got " + value.Type().String())
+	}
+	promVec := value.(pmodel.Vector)
+	for _, sample := range promVec {
+		metric := sample.Metric
+		bucketName := metric["le"] //hardcode bucket upper bound
+		CurrentSQBInfo[string(bucketName)] = int(sample.Value)
+	}
+	return nil
+}
+
+func querySlowQueryMetric(sctx sessionctx.Context) (result pmodel.Value, err error) {
+	// Add retry to avoid network error.
+	var prometheusAddr string
+	for i := 0; i < 5; i++ {
+		//TODO: the prometheus will be Integrated into the PD, then we need to query the prometheus in PD directly, which need change the quire API
+		prometheusAddr, err = infosync.GetPrometheusAddr()
+		if err == nil || err == infosync.ErrPrometheusAddrIsNotSet {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	if err != nil {
+		return nil, err
+	}
+	promClient, err := api.NewClient(api.Config{
+		Address: prometheusAddr,
+	})
+	if err != nil {
+		return nil, err
+	}
+	promQLAPI := promv1.NewAPI(promClient)
+	promQL := "tidb_server_slow_query_process_duration_seconds_bucket{sql_type=\"general\"}"
+
+	ts := time.Now()
+	// Add retry to avoid network error.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	for i := 0; i < 5; i++ {
+		result, _, err = promQLAPI.Query(ctx, promQL, ts)
+		if err == nil {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	return result, err
+}
+
+// calculateDeltaSQB calculate the delta between current slow query bucket and last slow query bucket
+func calculateDeltaSQB() *SlowQueryBucket {
+	deltaMap := make(SlowQueryBucket)
+	for key, value := range CurrentSQBInfo {
+		deltaMap[key] = value - (LastSQBInfo)[key]
+	}
+	return &deltaMap
+}
+
+// initSlowQueryStats Init LastSQBInfo, follow the definition of metrics/server.go
+// Buckets:   prometheus.ExponentialBuckets(0.001, 2, 28), // 1ms ~ 1.5days
+func initSlowQueryStats() {
+	LastSQBInfo := make(SlowQueryBucket)
+	CurrentSQBInfo := make(SlowQueryBucket)
+
+	bucketBase := 0.001 // From 0.001 to 134217.728, total 28 float number; the 29th is +Inf
+	for i := 0; i < slowQueryBucketNum-1; i++ {
+		LastSQBInfo[strconv.FormatFloat(bucketBase, 'f', 3, 32)] = 0
+		CurrentSQBInfo[strconv.FormatFloat(bucketBase, 'f', 3, 32)] = 0
+		bucketBase += bucketBase
+	}
+	LastSQBInfo["+Inf"] = 0
+	CurrentSQBInfo["+Inf"] = 0
+
+	logutil.BgLogger().Info("Telemetry slow query stats initialized", zap.String("CurrentSQBInfo", CurrentSQBInfo.String()))
```

Comment:
You can merge two log into one:

```go
logutil.BgLogger().Info("Telemetry slow query stats initialized", zap.String("CurrentSQBInfo", CurrentSQBInfo.String()), zap.String("LastSQBInfo", LastSQBInfo.String()))
```

Reply:
......., such a stupid code......, Accepted.